Optimize DicomTag.GetHashCode

amoerie opened the pull request:
Fixes #1088 .

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [ ] ~~I have updated API documentation~~
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Do not use ToString() in DicomTag.GetHashCode
- Reset cached hash value when setting Private Creator (this field can be publicly set..)
- Add suite of tests that verify we don't produce hash collisions for known DICOM tags
- Fix Service Provider resolution in unit tests when debugging. Some stack frames do not have namespaces, causing null reference exceptions

amoerie commented:
I reverted "Dispose ImageSharp image when disposing the Fellow Oak DICOM image", that commit should not have been there, sorry.

amoerie commented:
Looks like the only failing unit test is one related to the codecs again.

gofal commented on Tests/Desktop/DicomTagTest.cs:
```diff
@@ -58,6 +59,118 @@ public void Contains_SamePrivateTagsDifferentPrivateCreator_ReturnsTrue()
+        {
+            var hashCode = DicomTag.AccessionNumber.GetHashCode();
+
+            Assert.InRange(hashCode, 1, int.MaxValue);
```
this may fail for good reason. Hashcodes have to be integers, but also may be netagive values. So `PrivateCreator.GetHashCode()` may return some negative value anytime.

see here in microsoft doc, there are examples with negative hash-values. https://docs.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=netcore-3.1#examples

amoerie replied:
You're right. I forgot to change this unit test to verify that it's not zero, that's the only "invalid value" AFAIK. I'll modify when I get the chance.

amoerie commented:
I removed the unit test for both v4 and v5. I'm not 100% sure 0 is not a valid hash code, the main purpose of hash codes is to distinguish between objects, so perhaps it's OK. The other unit tests should suffice in validating the hashcode calculation. 

I also removed the cached _hash value, as you showed that it's not really valuable to do so.
